library/core/tests/num/mod.rs

```diff
@@ -870,6 +870,17 @@ macro_rules! test_float {
                 assert!(($nan as $fty).midpoint(1.0).is_nan());
                 assert!((1.0 as $fty).midpoint($nan).is_nan());
                 assert!(($nan as $fty).midpoint($nan).is_nan());
+
+                // test if large differences in magnitude are still correctly computed.
+                // NOTE: that because of how small x and y are, x + y can never overflow
+                // so (x + y) / 2.0 is always correct
+                for i in 64..128 {
+                    for j in 0..64u8 {
+                        let x = (2.0f32.powi(i) + f32::from(j)) / 2.0;
+                        let y = 2.0f32.powi(i).midpoint(f32::from(j));
+                        assert_eq!(x, y);
+                    }
+                }
```

Comment:
Ah another thing, the testsuite is supposed to be generic over f32/f64 but this only does f32 things. Either it should be a separate test outside the macro or it should work for f64 too.

And the x86 tests failed on f64 too, so there probably is an issue with the test, not just the midpoint calculation. So it might still fail even when using the old implementation.

You might want to test this locally by installing 32bit glibc and building for an i586 target (running the docker CI scripts).

Reply:
Ok, I tested this on `i686-gnu` via docker, and that worked fine.
I've also extended the tests to include the limits of `f64` as well (and portable to `f16` and `f128` when those get included).